### PR TITLE
Mt8186 pcm

### PR DESCRIPTION
**Pull request by chunxu-li**
Hi ,
This series introduce the support of MT8186 SOF audio driver.

### CODE REVIEW

File: sound/soc/sof/mediatek/mt8186/mt8186.c [hunk 1]
```diff
@@ -27,6 +27,12 @@
 #include "mt8186.h"
 #include "mt8186-clk.h"
 
+struct snd_sof_mach_info {
```
Comment (dbaluta) [inline sound/soc/sof/mediatek/mt8186/mt8186.c]: A patch should mostly do one logical thing not more. Your does a bunch of unrelated stuff.

I would create multiple patches like this:

1) add snd_sof_dsp_ops callbacks for sof_xtensa_arch_ops / open / close / mailbox_read / mailbox_write (this could be all added in 1 patch).

2) change of ipc_msg_data and explain why

3) adding of `mt8186_dai`

4) add the machine_info_select and related code.

Comment (chunxu-li) [reply sound/soc/sof/mediatek/mt8186/mt8186.c]: Thanks for reminder， I'll split the code.

File: sound/soc/sof/mediatek/mt8186/mt8186.c [hunk 2]
```diff
@@ -460,14 +466,59 @@ static int mt8186_get_bar_index(struct snd_sof_dev *sdev, u32 type)
...
-			       void *p, size_t sz)
+static void *mt8186_machine_info_select(struct snd_sof_dev *sdev)
 {
-	sof_mailbox_read(sdev, sdev->dsp_box.offset, p, sz);
-	return 0;
+	struct snd_sof_pdata *sof_pdata = sdev->pdata;
+	const struct sof_dev_desc *desc = sof_pdata->desc;
+	struct snd_sof_mach_info *mach_info = (struct snd_sof_mach_info *)desc->mach_info;
+
+	for (; mach_info->board; mach_info++) {
```
Comment (yaochunhung) [inline sound/soc/sof/mediatek/mt8186/mt8186.c]: Check if mach_info is NULL before use it. Thanks

File: sound/soc/sof/mediatek/mt8186/mt8186.c [hunk 3]
```diff
@@ -460,14 +466,59 @@ static int mt8186_get_bar_index(struct snd_sof_dev *sdev, u32 type)
...
+		int score = of_machine_is_compatible(mach_info->board);
+		dev_dbg(sdev->dev, "%s, score:%d, board:%s\n",
```
Comment (yaochunhung) [inline sound/soc/sof/mediatek/mt8186/mt8186.c]: Add one blank line between local declaration and statement. Thanks.

File: sound/soc/sof/mediatek/mt8186/mt8186.c (hunk 2 above)
Comment (chunxu-li) [reply sound/soc/sof/mediatek/mt8186/mt8186.c]: done

File: sound/soc/sof/mediatek/mt8186/mt8186.c (hunk 3 above)
Comment (chunxu-li) [reply sound/soc/sof/mediatek/mt8186/mt8186.c]: done

File: include/sound/sof.h [hunk 4]
```diff
@@ -89,6 +89,7 @@ struct snd_sof_pdata {
 	/* machine */
 	struct platform_device *pdev_mach;
 	const struct snd_soc_acpi_mach *machine;
+	const void *mach_info;
```
Comment (ranj063) [inline include/sound/sof.h]: @dbaluta out of curiosity how do you handle for the IMX platforms?

**Review by kuanhsuncheng — APPROVED** [review]

**Review by yaochunhung — APPROVED** [review]

File: include/sound/sof.h (hunk 4 above)
Comment (dbaluta) [reply include/sound/sof.h]: @ranj063, in our internal tree we read this info from device tree but in a very non standard way.

https://www.spinics.net/lists/devicetree/msg433908.html

We got lots of advice from community on how to handle this, but after several iterations i got busy with other stuff. I think @chunxu-li is the way to go. Will update with similar patches for i.MX

Comment (ranj063) [reply include/sound/sof.h]: Thanks @dbaluta this is what I wanted to know i fthis will be used in other platforms as well.

File: sound/soc/sof/mediatek/mt8186/mt8186.c (hunk 1 above)
Comment (ranj063) [inline sound/soc/sof/mediatek/mt8186/mt8186.c]: @chunxu-li since there are plans for others to use the same methodology, do you think this struct definition should be moved outsidemaybe?

Comment (ranj063) [reply sound/soc/sof/mediatek/mt8186/mt8186.c]: and maybe rename this to be snd_sof_of_mach to be consistent with acpi_mach.

Comment (chunxu-li) [reply sound/soc/sof/mediatek/mt8186/mt8186.c]: @ranj063  Thanks a lot for your advice. In the current sof file structure, which header do you think is more appropriate to put this declaration?

File: sound/soc/sof/pcm.c [hunk 5]
```diff
@@ -654,7 +654,10 @@ void snd_sof_new_platform_drv(struct snd_sof_dev *sdev)
 	struct snd_sof_pdata *plat_data = sdev->pdata;
 	const char *drv_name;
 
-	drv_name = plat_data->machine->drv_name;
+	if (plat_data->machine)
+		drv_name = plat_data->machine->drv_name;
+	else
+		drv_name = NULL;
```
Comment (ujfalusi) [inline sound/soc/sof/pcm.c:663]: Should somehow the `pd->ignore_machine` initialized to something not NULL for non ACPI platforms?

File: sound/soc/sof/mediatek/mt8186/mt8186.c (hunk 1 above)
Comment (ujfalusi) [reply sound/soc/sof/mediatek/mt8186/mt8186.c]: @chunxu-li, I would probably move the struct definition to `sound/soc/sof/sof-priv.h` and the `mt8186_machine_info_select()` to `sound/soc/sof/sof-of-dev.c` (+.h) as `sof_of_machine_info_select()`
Or move the struct definition to sof-of-dev.h as well since it is used for OF only?

File: sound/soc/sof/pcm.c (hunk 5 above)
Comment (chunxu-li) [reply sound/soc/sof/pcm.c:663]: Hi @ujfalusi , current patch can resolve [this pull request as well](https://github.com/thesofproject/linux/pull/3352).

The 'ignore_machine' field is currently used to ignore all FE dailinks statically added by the machine drivers, as well as override the fixups for the BE dailinks. The motivation for this field was primarily to reuse the same machine driver on Intel devices, both with legacy and SOF-based platform drivers.
The other platforms, where the same card may uses SOF-based dailinks to deal with DSP-managed streams, as well as 'regular' dailinks. The 'ignore_machine' field set by the core SOF platform driver is too strong, with dailinks not managed by SOF being modified. 

update as
```	
        if (plat_data->machine)
		drv_name = plat_data->machine->drv_name;
	else if (plat_data->of_machine)
		drv_name = plat_data->of_machine->drv_name;
	else
		drv_name = NULL;
```

File: sound/soc/sof/mediatek/mt8186/mt8186.c (hunk 1 above)
Comment (kuanhsuncheng) [reply sound/soc/sof/mediatek/mt8186/mt8186.c]: @dbaluta 

I think it is ok to move this struct definition to sof-of-dev.h. What do you feel about this idea?

File: sound/soc/sof/sof-of-dev.h [hunk 6]
```diff
@@ -9,8 +9,16 @@
 #ifndef __SOUND_SOC_SOF_OF_H
 #define __SOUND_SOC_SOF_OF_H
 
+struct snd_sof_of_mach {
+	const char *board;
+	const char *drv_name;
+	const char *fw_filename;
+	const char *sof_tplg_filename;
```
Comment (chunxu-li) [inline sound/soc/sof/sof-of-dev.h:16]: Hi @dbaluta  I checked this link [https://www.spinics.net/lists/devicetree/msg433908.html](https://www.spinics.net/lists/devicetree/msg433908.html)
the following members should meet your requirement,  do you have any other advice?
`drv_name`
`tplg_filename`
`fw_filename`

File: sound/soc/sof/sof-priv.h [hunk 7]
```diff
@@ -283,6 +283,7 @@ struct snd_sof_dsp_ops {
 	void (*machine_unregister)(struct snd_sof_dev *sdev,
 				   void *pdata); /* optional */
 	struct snd_soc_acpi_mach * (*machine_select)(struct snd_sof_dev *sdev); /* optional */
+	struct snd_sof_of_mach * (*machine_info_select)(struct snd_sof_dev *sdev); /* optional */
```
Comment (ujfalusi) [inline sound/soc/sof/sof-priv.h]: can we name this as `of_machine_select`?

In a followup we could rename the existing `machine_select` to `acpi_machine_select` and then modify the wrapper in sound/soc/sof/ops.h to:
```c
static inline void *snd_sof_machine_select(struct snd_sof_dev *sdev)
{
	if (!sof_ops(sdev))
		return NULL;

	if (sof_ops(sdev)->acpi_machine_select)
		return sof_ops(sdev)->acpi_machine_select(sdev);
	else if (sof_ops(sdev)->of_machine_select)
		return sof_ops(sdev)->of_machine_select(sdev);

	return NULL;
}
```
It might make things a bit easier to read?

Comment (chunxu-li) [reply sound/soc/sof/sof-priv.h]: Hi @ujfalusi very appreciate for your advice, I'll update this part of code as soon as possible

File: sound/soc/sof/sof-of-dev.h (hunk 6 above)
Comment (tinghan-shen) [inline sound/soc/sof/sof-of-dev.h:16]: Maybe use "tplg_filename" instead to align with snd_sof_acpi_mach?

File: sound/soc/sof/sof-of-dev.c [hunk 8]
```diff
@@ -31,6 +31,35 @@ const struct dev_pm_ops sof_of_pm = {
...
+	for (; mach->board; mach++) {
+		int score = of_machine_is_compatible(mach->board);
+
+		dev_dbg(sdev->dev, "%s, score:%d, board:%s\n",
+			__func__, score, mach->board);
+		if (score) {
+			sof_pdata->tplg_filename = mach->sof_tplg_filename;
+			if (mach->fw_filename)
+				sof_pdata->fw_filename = mach->fw_filename;
+			dev_info(sdev->dev, "%s ,tplg:%s, fw:%s\n",
```
Comment (tinghan-shen) [inline sound/soc/sof/sof-of-dev.c]: This appears to be a debug message. Should it be removed or rewritten as a sentence?

File: sound/soc/sof/sof-of-dev.c [hunk 9]
```diff
@@ -31,6 +31,35 @@ const struct dev_pm_ops sof_of_pm = {
...
+{
+	struct snd_sof_pdata *sof_pdata = sdev->pdata;
+	const struct sof_dev_desc *desc = sof_pdata->desc;
+	struct snd_sof_of_mach *mach = desc->of_machines;
+
+	if (!mach)
+		return NULL;
+
+	for (; mach->board; mach++) {
+		int score = of_machine_is_compatible(mach->board);
```
Comment (tinghan-shen) [inline sound/soc/sof/sof-of-dev.c]: The 'score' looks like for debug purpose. May be we can change it to following?

if (of_machine_is_compatible(mach->board)) {
 // ...
}

Comment (chunxu-li) [reply sound/soc/sof/sof-of-dev.c]: Done

File: sound/soc/sof/sof-of-dev.c (hunk 8 above)
Comment (chunxu-li) [reply sound/soc/sof/sof-of-dev.c]: Done

Comment (afq984) [reply sound/soc/sof/sof-of-dev.c]: cc @Zames-Chang

File: sound/soc/sof/ops.h [hunk 10]
```diff
@@ -522,6 +522,15 @@ snd_sof_set_mach_params(struct snd_soc_acpi_mach *mach,
 		sof_ops(sdev)->set_mach_params(mach, sdev);
 }
 
+static inline struct snd_sof_of_mach *
+snd_sof_of_machine_select(struct snd_sof_dev *sdev)
+{
+	if (sof_ops(sdev) && sof_ops(sdev)->of_machine_select)
+		return sof_ops(sdev)->of_machine_select(sdev);
```
Comment (plbossart) [inline sound/soc/sof/ops.h:529]: I don't understand the definitions at all. In the DT world, the machine is  already described with a device, which will be matched with a driver already. What does 'of_machine_select' mean in this context?

File: include/sound/sof.h [hunk 11]
```diff
@@ -89,6 +89,7 @@ struct snd_sof_pdata {
 	/* machine */
 	struct platform_device *pdev_mach;
 	const struct snd_soc_acpi_mach *machine;
+	const struct snd_sof_of_mach *of_machine;
```
Comment (plbossart) [inline include/sound/sof.h:92]: grammar in commit message

a chance for the machines which are not based on ACPI, this callback
is optional, each machine that is not based on ACPI can implement

Comment (ranj063) [reply include/sound/sof.h:92]: @plbossart do you think it makes better sense to make the machine field const void * in snd_sof_pdata and resolve it to either snd_soc_acpi_mach or and_sof_of_mach in the platform dependent code?

Comment (plbossart) [reply include/sound/sof.h:92]: possibly, but it's not clear to me by the DT/OF machines need something in the kernel when everything is already described in platform firmware. That's the big difference between DT/OF and ACPI, in ACPI we do not have a device entry for the machine driver, so we do all this machine driver selection. It's not needed for DT/OF, so it's really questionable why we would build on something that was a work-around ACPI limitations.

Comment (chunxu-li) [reply include/sound/sof.h:92]: @plbossart @ranj063 , Sorry for confusion.
If we do not implement the machine_select callback the sof_machine_check will return -ENODEV. what we actually want is do not return error since we don't need machine_select with DT/OF machines. What's more the sof_machine_select not just checking ACPI machine, but also assign sof_tplg_filename at the same time. Do you have any ideal how to handle this?


```
int sof_machine_check(struct snd_sof_dev *sdev)
{
//...
	if (!IS_ENABLED(CONFIG_SND_SOC_SOF_FORCE_NOCODEC_MODE)) {

		/* return ACPI machine, this will return NULL with DT/OF machines */
		mach = snd_sof_machine_select(sdev); 
		if (mach) {
			sof_pdata->machine = mach;
			snd_sof_set_mach_params(mach, sdev);
			return 0;
		}

                /* this will lead to machine check fail. */
		if (!IS_ENABLED(CONFIG_SND_SOC_SOF_NOCODEC)) {
			dev_err(sdev->dev, "error: no matching ASoC machine driver found - aborting probe\n");
			return -ENODEV; 
		}
// ...
}
```

Comment (ranj063) [reply include/sound/sof.h:92]: @plbossart that is indeed a valid question. We need to find the ACPI machine and register it in sof_machine_register(). Without the need to explicitly register the machine driver, I cant explain what the need for change for OF devices would be for. @dbaluta any ideas?

Comment (chunxu-li) [reply include/sound/sof.h:92]: Hi @ranj063  @plbossart @dbaluta The 'ACPI machine' should be selected before register by `sof_machine_register`, then it looks like the `machine_select` and `machine_register` should use at the same time when using ACPI machine.

How about
1. we add a condition before `machine_select`? 
or
2. add a flag to check whether the current machine is ACPI or not? 


1.check whether the machine_register is implement or not, if the `machine_register` is implemented then the machine must be checked before register.

```
int sof_machine_check(struct snd_sof_dev *sdev)
{
//...
	if (!IS_ENABLED(CONFIG_SND_SOC_SOF_FORCE_NOCODEC_MODE)) {
                 /* acpi machine should implement machine register */
                if (sof_ops(dev) && sof_ops(dev)->machine_register) {
                        /* return ACPI machine */
		        mach = snd_sof_machine_select(sdev); 
        		if (mach) {
	        		sof_pdata->machine = mach;
		        	snd_sof_set_mach_params(mach, sdev);
			        return 0;
		        }
                } else {
                        /* of machine device return directly */
                        return 0;
                }

                /* this will lead to machine check fail. */
		if (!IS_ENABLED(CONFIG_SND_SOC_SOF_NOCODEC)) {
			dev_err(sdev->dev, "error: no matching ASoC machine driver found - aborting probe\n");
			return -ENODEV; 
		}
// ...
}
```

2. add a flag  in sof_dev_desc to check whether the current machine is ACPI or not? 
```
int sof_machine_check(struct snd_sof_dev *sdev)
{
//...
        const struct sof_dev_desc *desc = sof_pdata->desc;
	if (!IS_ENABLED(CONFIG_SND_SOC_SOF_FORCE_NOCODEC_MODE)) {
                 /* acpi machine should select machine before register */
                if (desc->is_acpi_machine) {
                        /* return ACPI machine */
		        mach = snd_sof_machine_select(sdev); 
        		if (mach) {
	        		sof_pdata->machine = mach;
		        	snd_sof_set_mach_params(mach, sdev);
			        return 0;
		        }
                } else {
                        /* of machine device return directly */
                        return 0;
                }

                /* this will lead to machine check fail. */
		if (!IS_ENABLED(CONFIG_SND_SOC_SOF_NOCODEC)) {
			dev_err(sdev->dev, "error: no matching ASoC machine driver found - aborting probe\n");
			return -ENODEV; 
		}
// ...
}
```

**Comment by kuanhsuncheng** [comment]
It seems there is no further discussion. Could someone please give us some comments? thanks.

File: sound/soc/sof/sof-of-dev.h (hunk 6 above)
Comment (dbaluta) [reply sound/soc/sof/sof-of-dev.h:16]: @chunxu-li what I don't understand from your patch is where do we get for example `drv_name`. 

Mark, here https://www.spinics.net/lists/devicetree/msg433913.html advised that this should be taken from something like `sof_dev_desc`.

Comment (chunxu-li) [reply sound/soc/sof/sof-of-dev.h:16]: Hi @dbaluta :

_this should be taken from something like sof_dev_desc._ 
Yes,  We add this mach info into  sof_of_mt8186_desc.of_machines field, and assign `drv_name` in function `snd_sof_new_platform_drv` at sof/pcm.c 
code example
```
static struct snd_sof_of_mach sof_mt8186_machs[] = {
	{
		.board = "google,kingler",
		.sof_tplg_filename = "sof-mt8186-mt6366-da7219-max98357.tplg",
                .drv_name = "mt8186-mt6366-da7219-max98357"
	},
	{
		.board = "google,krabby",
		.sof_tplg_filename = "sof-mt8186-mt6366-rt1019-rt5682s.tplg",
                .drv_name = "mt8186-mt6366-rt1019-rt5682s"
	},
	{}
};

static const struct sof_dev_desc sof_of_mt8186_desc = {
	.of_machines = sof_mt8186_machs,
```

```
// sof/pcm.c  
// void snd_sof_new_platform_drv(struct snd_sof_dev *sdev)
        if (plat_data->machine)
		drv_name = plat_data->machine->drv_name;
	else if (plat_data->of_machine)
		drv_name = plat_data->of_machine->drv_name;
```

Comment (dbaluta) [reply sound/soc/sof/sof-of-dev.h:16]: I see. This works for me.

File: sound/soc/sof/ops.h (hunk 10 above)
Comment (dbaluta) [reply sound/soc/sof/ops.h:529]: do you mean 'a machine is already described with a device using device tree, right?'. Device tree describes the hardware, while we need information like machine driver name, tplg name, firmware name which according to Mark cannot be placed inside a DT node. 

https://www.spinics.net/lists/devicetree/msg433913.html

**Review by dbaluta — APPROVED** [review]

**Comment by kuanhsuncheng** [comment]
HI @plbossart @ranj063  @ujfalusi 

Please let us know if there is anything chunxu can improve on, Thanks.

**Review by kuanhsuncheng — APPROVED** [review]

**Review by yaochunhung — APPROVED** [review]

File: include/sound/sof.h (hunk 11 above)
Comment (kuanhsuncheng) [reply include/sound/sof.h:92]: @plbossart, 
From devicetree people comment, it's not ok we put fw/tplg filenames in DT.  I think ACPI and DT are different specifications and it's not a work-around.

We need find a way for sof_machine_check func. 

Please let me know if you have any comments. thanks.

**Comment by Zames-Chang** [comment]
Ping.  We think we addressed all comments, is there anything else that needs to change?
